src/main/java/org/jvnet/hudson/test/LoggerRule.java

```diff
@@ -0,0 +1,163 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jvnet.hudson.test;
+
+import hudson.util.RingBufferLogHandler;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.ConsoleHandler;
+import java.util.logging.Formatter;
+import java.util.logging.Level;
+import java.util.logging.LogRecord;
+import java.util.logging.Logger;
+import java.util.logging.SimpleFormatter;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.rules.ExternalResource;
+import org.junit.rules.RuleChain;
+
+/**
+ * A test rule which allows you to easily enable one or more loggers for the duration of a test.
+ * Call {@link #record(Class, Level)} or another overload for the rule to take effect.
+ * <p>By default messages are merely printed to test output.
+ * If you also want to examine them, call TODO
+ * <p>To capture messages during Jenkins startup,
+ * you may compose this with a {@link JenkinsRule} using a {@link RuleChain};
+ * or use as a {@link ClassRule}.
+ */
+public class LoggerRule extends ExternalResource {
+
+    private final ConsoleHandler consoleHandler = new ConsoleHandler();
+    private final Map<Logger,Level> loggers = new HashMap<Logger,Level>();
+    // initialized iff capture is called:
+    private RingBufferLogHandler ringHandler;
+    private List<String> messages;
+
+    /**
+     * Initializes the rule, by default not recording anything.
+     */
+    public LoggerRule() {
+        consoleHandler.setLevel(Level.ALL);
+    }
+
+    /**
+     * Initializes log record capture, in addition to merely printing it.
+     * This allows you to call {@link #getRecords} and/or {@link #getRecordsFormatted} later.
+     * @param maximum the maximum number of records to keep (any further will be discarded)
+     * @return this rule, for convenience
+     */
+    public LoggerRule capture(int maximum) {
+        messages = new ArrayList<String>();
+        ringHandler = new RingBufferLogHandler(maximum) {
+            final Formatter f = new SimpleFormatter(); // placeholder instance for what should have been a static method perhaps
+            @Override
+            public synchronized void publish(LogRecord record) {
+                super.publish(record);
+                String message = f.formatMessage(record);
+                Throwable x = record.getThrown();
+                messages.add(message == null && x != null ? x.toString() : message);
```

Comment:
> I take it the unbounded nature is what you are expecting here.

Did I miss something that was limiting the number of messages kept in the `ArrayList` rather than the ring buffer?